### PR TITLE
connection: tighten network connection buffer limits

### PR DESCRIPTION
**Pull request by antoniovicente**
Commit Message:
connection: tighten network connection buffer limits

Risk Level: low. small change to edge case threshold, minimal functional changes expected.
Testing: extended unit tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

**Comment by antoniovicente** [comment]
merged master to fix the coverage failure.

File: source/common/network/connection_impl.cc
Comment (antoniovicente) [inline source/common/network/connection_impl.cc:492]: I should mention that the +1 on the limit can result in a uint32_t overflow. One example where this happens is source/extensions/filters/network/dubbo_proxy/conn_manager.cc which passes in UINT32_MAX to setBufferLimits

**Comment by antoniovicente** [comment]
Happy new year.

@alyssawilk @ggreenway  could you please take a look when you have a chance?

File: test/common/network/connection_impl_test.cc
Comment (ggreenway) [inline test/common/network/connection_impl_test.cc:986]: "2 bytes"

**Review by ggreenway — CHANGES_REQUESTED** [review]
This looks good, modulo one comment fix. I really like the additional test coverage around the edge cases.

I'd like @alyssawilk to take a look before merging as she's much more familiar with the watermarking code.

**Review by alyssawilk — COMMENTED** [review]
code LGTM, though I would strongly encourage runtime guards in case say this exposes bugs elsewhere in the code base.  I'll let @ggreenway tie break as I tend to err very far on the side of caution.

Either way that goes, I think there should be a release note

**Comment by antoniovicente** [comment]
> code LGTM, though I would strongly encourage runtime guards in case say this exposes bugs elsewhere in the code base. I'll let @ggreenway tie break as I tend to err very far on the side of caution.
> 
> Either way that goes, I think there should be a release note

The main reason for this change is to unblock https://github.com/envoyproxy/envoy/pull/14054, which depends on the exact value of this limit.  Thinking about it more, I think the implication of the off-by-1 in limits is that we'ld do an extra 1 byte read but we won't trigger high watermark like I thought we would.  Triggering high watermark was the more concerning of the efficiency issues.

I don't know if runtime guarding of this PR implies that https://github.com/envoyproxy/envoy/pull/14054 should also be runtime guarded. The change in that PR is less subtle than this one, but partially depends on this one and seems more complex than this one.

I'm struggling a bit with what I should say in the release note, will get back to it ASAP.

**Comment by alyssawilk** [comment]
haven't looked at the other PR.  If you can guard both, that may be an option but I can trust you on if it'd be more risky than not given the other PR.

**Comment by ggreenway** [comment]
I think I'm ok with no runtime guard, because there are tests validating the behavior is correct, and it would only be a perf issue if we get it wrong, not a functional issue. But I don't feel strongly about it.

**Comment by antoniovicente** [comment]
> code LGTM, though I would strongly encourage runtime guards in case say this exposes bugs elsewhere in the code base. I'll let @ggreenway tie break as I tend to err very far on the side of caution.
> 
> Either way that goes, I think there should be a release note

Added release note.

Sorry for the delay.  PTAL.

**Review by alyssawilk — APPROVED** [review]

**Review by ggreenway — APPROVED** [review]

**Pull request merged**
